File: backend/test/games.e2e.test.ts
import { afterAll, beforeAll, beforeEach, describe, expect, it } from 'vitest';
import request from 'supertest';

import { computeRoundScore, haversineKm } from '../src/games/scoring.js';
import type { GameMode } from '../src/games/types.js';
import {
    closeInfra,
    getApp,
    makeRandomEmail,
    obtainVerificationCode,
    prepareDatabase,
    resetAuthState,
} from './helpers.js';

const VALID_PASSWORD = 'secret123';

// 每轮：给定答案/猜测坐标，用服务端同款公式生成必然一致的得分，保证提交通过校验
const ANSWERS = [
    { lat: 39.9055, lng: 116.3976 }, // 北京
    { lat: 31.2304, lng: 121.4737 }, // 上海
    { lat: 23.1291, lng: 113.2644 }, // 广州
];
const GUESS_OFFSETS = [
    { lat: 0.01, lng: 0.01 },
    { lat: 0.05, lng: -0.05 },
    { lat: 0.002, lng: 0.002 },
];
const DEFAULT_ANSWER = ANSWERS[0]!;
const DEFAULT_OFFSET = GUESS_OFFSETS[0]!;

function makeRound(mode: GameMode, region: string | null, index: number): Record<string, unknown> {
    const answer = ANSWERS[index] ?? DEFAULT_ANSWER;
    const offset = GUESS_OFFSETS[index] ?? DEFAULT_OFFSET;
    const guess = { lat: answer.lat + offset.lat, lng: answer.lng + offset.lng };
    const distanceKm = haversineKm(answer.lat, answer.lng, guess.lat, guess.lng);
    return {
        name: `测试点${index + 1}`,
        distanceKm,
        score: computeRoundScore(mode, region, distanceKm),
        imageId: `img-${index}`,
        xp: 0,
        difficulty: 3,
        guessLat: guess.lat,
        guessLng: guess.lng,
        answerLat: answer.lat,
        answerLng: answer.lng,
    };
}

function makeTimeoutRound(): Record<string, unknown> {
    const answer = ANSWERS[2] ?? DEFAULT_ANSWER;
    return {
        name: '超时点位',
        distanceKm: null,
        score: 0,
        imageId: null,
        xp: 0,
        difficulty: 2,
        guessLat: null,
        guessLng: null,
        answerLat: answer.lat,
        answerLng: answer.lng,
    };
}

function makeGame(
    mode: GameMode,
    region: string | null,
    roundCount: number,
    withTimeout = false
): { totalScore: number; rounds: Array<Record<string, unknown>> } {
    const rounds = Array.from({ length: roundCount }, (_, index) => makeRound(mode, region, index));
    if (withTimeout) {
        rounds.push(makeTimeoutRound());
    }
    const totalScore = rounds.reduce((sum, round) => sum + (round.score as number), 0);
    return { totalScore, rounds };
}

async function createGuestToken(): Promise<string> {
    const response = await request(getApp()).post('/api/auth/guest');
    expect(response.status).toBe(201);
    return response.body.guestToken as string;
}

async function createUserSession(): Promise<{ email: string; accessToken: string; guestToken?: string }> {
    const email = makeRandomEmail('games');
    const code = await obtainVerificationCode(email);
    const response = await request(getApp())
        .post('/api/auth/register')
        .send({ username: 'gamer01', email, password: VALID_PASSWORD, code });
    expect(response.status).toBe(201);
    return { email, accessToken: response.body.tokenPair.accessToken as string };
}

function submitGame(token: string, body: Record<string, unknown>) {
    return request(getApp()).post('/api/games').set('Authorization', `Bearer ${token}`).send(body);
}

beforeAll(async () => {
    await prepareDatabase();
});

beforeEach(async () => {
    await resetAuthState();
});

afterAll(async () => {
    await closeInfra();
});

describe('POST /api/games 提交成绩', () => {
    it('游客提交成绩成功并写入回合明细', async () => {
        const token = await createGuestToken();
        const game = makeGame('challenge', null, 3, true);
        const response = await submitGame(token, {
            mode: 'challenge',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });
        expect(response.status).toBe(201);
        expect(response.body.game.id).toBeTruthy();
        expect(response.body.game.rounds).toHaveLength(4);
        expect(response.body.game.totalScore).toBe(game.totalScore);
    });

    it('未登录返回 401', async () => {
        const game = makeGame('classic', null, 1);
        const response = await request(getApp())
            .post('/api/games')
            .send({ mode: 'classic', totalScore: game.totalScore, rounds: game.rounds });
        expect(response.status).toBe(401);
    });

    it('总分与回合明细不一致返回 400', async () => {
        const token = await createGuestToken();
        const game = makeGame('classic', null, 1);
        const response = await submitGame(token, {
            mode: 'classic',
            totalScore: 9999,
            rounds: game.rounds,
        });
        expect(response.status).toBe(400);
    });

    it('单轮得分与距离不符返回 400', async () => {
        const token = await createGuestToken();
        const game = makeGame('classic', null, 1);
        const forgedRounds = [{ ...game.rounds[0], score: 5000 }];
        const response = await submitGame(token, {
            mode: 'classic',
            totalScore: 5000,
            rounds: forgedRounds,
        });
        expect(response.status).toBe(400);
    });

    it('距离与提交坐标不符返回 400', async () => {
        const token = await createGuestToken();
        const game = makeGame('classic', null, 1);
        const forgedDistance = 0.1;
        const forgedRounds = [
            {
                ...game.rounds[0],
                distanceKm: forgedDistance,
                score: computeRoundScore('classic', null, forgedDistance),
            },
        ];
        const response = await submitGame(token, {
            mode: 'classic',
            totalScore: forgedRounds[0]!.score as number,
            rounds: forgedRounds,
        });
        expect(response.status).toBe(400);
    });

    it('非法模式返回 400', async () => {
        const token = await createGuestToken();
        const game = makeGame('classic', null, 1);
        const response = await submitGame(token, {
            mode: 'unknown',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });
        expect(response.status).toBe(400);
    });

    it('空回合列表返回 400', async () => {
        const token = await createGuestToken();
        const response = await submitGame(token, { mode: 'classic', totalScore: 0, rounds: [] });
        expect(response.status).toBe(400);
    });

    it('单轮得分超限返回 400', async () => {
        const token = await createGuestToken();
        const response = await submitGame(token, {
            mode: 'classic',
            totalScore: 99999,
            rounds: [{ name: '北京', distanceKm: 0.1, score: 99999 }],
        });
        expect(response.status).toBe(400);
    });

    it('区域模式缺少 region 返回 400', async () => {
        const token = await createGuestToken();
        const game = makeGame('region', 'asia', 1);
        const response = await submitGame(token, {
            mode: 'region',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });
        expect(response.status).toBe(400);
    });

    it('非区域模式携带 region 返回 400', async () => {
        const token = await createGuestToken();
        const game = makeGame('classic', null, 1);
        const response = await submitGame(token, {
            mode: 'classic',
            region: 'asia',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });
        expect(response.status).toBe(400);
    });

    it('注册用户提交成绩成功', async () => {
        const session = await createUserSession();
        const game = makeGame('classic', null, 1);
        const response = await submitGame(session.accessToken, {
            mode: 'classic',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });
        expect(response.status).toBe(201);
    });
});

describe('GET /api/games/recent 历史记录', () => {
    it('按时间倒序返回最近记录，并受 limit 约束', async () => {
        const token = await createGuestToken();
        const classic = makeGame('classic', null, 1);
        const challenge = makeGame('challenge', null, 3, true);
        await submitGame(token, { mode: 'classic', totalScore: classic.totalScore, rounds: classic.rounds });
        await submitGame(token, { mode: 'challenge', totalScore: challenge.totalScore, rounds: challenge.rounds });

        const response = await request(getApp())
            .get('/api/games/recent?limit=1')
            .set('Authorization', `Bearer ${token}`);
        expect(response.status).toBe(200);
        expect(response.body.games).toHaveLength(1);
        expect(response.body.games[0].mode).toBe('challenge');
    });

    it('只能看到自己的记录', async () => {
        const tokenA = await createGuestToken();
        const tokenB = await createGuestToken();
        const classic = makeGame('classic', null, 1);
        await submitGame(tokenA, { mode: 'classic', totalScore: classic.totalScore, rounds: classic.rounds });

        const response = await request(getApp()).get('/api/games/recent').set('Authorization', `Bearer ${tokenB}`);
        expect(response.status).toBe(200);
        expect(response.body.games).toHaveLength(0);
    });

    it('无记录时返回空数组', async () => {
        const token = await createGuestToken();
        const response = await request(getApp()).get('/api/games/recent').set('Authorization', `Bearer ${token}`);
        expect(response.body.games).toEqual([]);
    });
});

describe('GET /api/games/best 最佳成绩', () => {
    it('返回该模式最高分记录', async () => {
        const token = await createGuestToken();
        const low = makeGame('classic', null, 1);
        const high = makeGame('classic', null, 2);
        await submitGame(token, { mode: 'classic', totalScore: low.totalScore, rounds: low.rounds });
        await submitGame(token, { mode: 'classic', totalScore: high.totalScore, rounds: high.rounds });

        const response = await request(getApp())
            .get('/api/games/best?mode=classic')
            .set('Authorization', `Bearer ${token}`);
        expect(response.status).toBe(200);
        expect(response.body.best.totalScore).toBe(high.totalScore);
    });

    it('不同模式互不干扰', async () => {
        const token = await createGuestToken();
        const classic = makeGame('classic', null, 1);
        const china = makeGame('china', null, 3, true);
        await submitGame(token, { mode: 'classic', totalScore: classic.totalScore, rounds: classic.rounds });
        await submitGame(token, { mode: 'china', totalScore: china.totalScore, rounds: china.rounds });

        const challengeResponse = await request(getApp())
            .get('/api/games/best?mode=challenge')
            .set('Authorization', `Bearer ${token}`);
        expect(challengeResponse.body.best).toBeNull();
    });

    it('无记录时返回 null', async () => {
        const token = await createGuestToken();
        const response = await request(getApp())
            .get('/api/games/best?mode=classic')
            .set('Authorization', `Bearer ${token}`);
        expect(response.body.best).toBeNull();
    });
});

describe('GET /api/games/summary 进度统计', () => {
    it('累计轮数/总分/最佳/猜中轮数', async () => {
        const token = await createGuestToken();
        const game = makeGame('challenge', null, 3, true);
        await submitGame(token, { mode: 'challenge', totalScore: game.totalScore, rounds: game.rounds });

        const response = await request(getApp()).get('/api/games/summary').set('Authorization', `Bearer ${token}`);
        expect(response.status).toBe(200);
        expect(response.body.progress).toMatchObject({
            totalRounds: 4,
            totalScore: game.totalScore,
            bestScore: game.totalScore,
            correctGuesses: 3,
        });
    });

    it('注册用户 /me 返回进度', async () => {
        const session = await createUserSession();
        const game = makeGame('classic', null, 1);
        await submitGame(session.accessToken, {
            mode: 'classic',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });

        const meResponse = await request(getApp())
            .get('/api/auth/me')
            .set('Authorization', `Bearer ${session.accessToken}`);
        expect(meResponse.body.progress).toMatchObject({ totalRounds: 1, bestScore: game.totalScore });
    });
});

describe('DELETE /api/games/:id 删除记录', () => {
    it('删除自己的记录成功', async () => {
        const token = await createGuestToken();
        const game = makeGame('classic', null, 1);
        const submitResponse = await submitGame(token, {
            mode: 'classic',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });
        const gameId = submitResponse.body.game.id as number;

        const deleteResponse = await request(getApp())
            .delete(`/api/games/${gameId}`)
            .set('Authorization', `Bearer ${token}`);
        expect(deleteResponse.status).toBe(200);

        const recentResponse = await request(getApp()).get('/api/games/recent').set('Authorization', `Bearer ${token}`);
        expect(recentResponse.body.games).toEqual([]);
    });

    it('删除他人记录返回 404', async () => {
        const tokenA = await createGuestToken();
        const tokenB = await createGuestToken();
        const game = makeGame('classic', null, 1);
        const submitResponse = await submitGame(tokenA, {
            mode: 'classic',
            totalScore: game.totalScore,
            rounds: game.rounds,
        });
        const gameId = submitResponse.body.game.id as number;

        const deleteResponse = await request(getApp())
            .delete(`/api/games/${gameId}`)
            .set('Authorization', `Bearer ${tokenB}`);
        expect(deleteResponse.status).toBe(404);
    });
});

File: backend/test/games_e2e_test.go
package e2e

import (
	"fmt"
	"net/http"
	"testing"

	"mma-guessr/backend/internal/games"
)

// buildRound builds a valid round payload for the given coordinates.
func buildRound(name string, guessLat, guessLng, answerLat, answerLng float64, distanceKm float64) map[string]any {
	return map[string]any{
		"name":       name,
		"locationId": nil,
		"distanceKm": distanceKm,
		"score":      0,
		"imageId":    nil,
		"xp":         0,
		"difficulty": 1,
		"guessLat":   guessLat,
		"guessLng":   guessLng,
		"answerLat":  answerLat,
		"answerLng":  answerLng,
	}
}

func TestLocations(t *testing.T) {
	e := newTestEnv(t)
	e.seedLocations(40)

	t.Run("random returns locations with count", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/locations/random?count=5", "", nil)
		if resp.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", resp.status)
		}
		locations := resp.nestedArray("locations")
		if len(locations) != 5 {
			t.Fatalf("expected 5 locations, got %d", len(locations))
		}
		first := locations[0].(map[string]any)
		for _, key := range []string{"id", "name", "lat", "lng", "region", "difficulty", "source"} {
			if _, ok := first[key]; !ok {
				t.Fatalf("location missing key %s: %v", key, first)
			}
		}
	})

	t.Run("random filters by region and difficulty", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/locations/random?region=asia&difficulty=1&count=10", "", nil)
		if resp.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", resp.status)
		}
		for _, raw := range resp.nestedArray("locations") {
			loc := raw.(map[string]any)
			if loc["region"] != "asia" || loc["difficulty"] != float64(1) {
				t.Fatalf("unexpected location %v", loc)
			}
		}
	})

	t.Run("stats returns totals", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/locations/stats", "", nil)
		if resp.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", resp.status)
		}
		if resp.body["total"] != float64(40) {
			t.Fatalf("expected total 40, got %v", resp.body["total"])
		}
		byRegion := resp.nestedMap("byRegion")
		if byRegion["asia"] != float64(33) || byRegion["europe"] != float64(7) {
			t.Fatalf("unexpected byRegion: %v", byRegion)
		}
	})

	t.Run("invalid count rejected", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/locations/random?count=999", "", nil)
		if resp.status != http.StatusBadRequest {
			t.Fatalf("expected 400, got %d", resp.status)
		}
	})
}

func TestGamesSubmit(t *testing.T) {
	e := newTestEnv(t)
	e.seedLocations(10)

	access := registerUser(t, e, uniqueUsername("tester"), randomEmail("games"), validPassword).
		nestedStr("tokenPair", "accessToken")

	// Exact known points: (30,100) and (31,101). Compute the authoritative
	// distance and score with the same formulas the server uses.
	const guessLat, guessLng, answerLat, answerLng = 30.0, 100.0, 31.0, 101.0
	exactDistance := games.HaversineKm(guessLat, guessLng, answerLat, answerLng)
	exactScore := games.ComputeRoundScore("classic", "", exactDistance)

	t.Run("submit valid game", func(t *testing.T) {
		round := buildRound("地点A", 30, 100, 31, 101, exactDistance)
		round["score"] = exactScore
		resp := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "classic",
			"totalScore": exactScore,
			"rounds":     []any{round},
		})
		if resp.status != http.StatusCreated {
			t.Fatalf("expected 201, got %d: %v", resp.status, resp.body)
		}
		game := resp.nestedMap("game")
		if game["totalScore"] != float64(exactScore) {
			t.Fatalf("unexpected totalScore %v", game["totalScore"])
		}
		rounds := resp.nestedArray("game", "rounds")
		if len(rounds) != 1 {
			t.Fatalf("expected 1 round, got %d", len(rounds))
		}
	})

	t.Run("recent returns submitted game", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/games/recent?limit=5", access, nil)
		if resp.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", resp.status)
		}
		games := resp.nestedArray("games")
		if len(games) != 1 {
			t.Fatalf("expected 1 game, got %d", len(games))
		}
	})

	t.Run("best returns top game", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/games/best?mode=classic", access, nil)
		if resp.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", resp.status)
		}
		if resp.nestedMap("best") == nil {
			t.Fatal("expected best game")
		}
	})

	t.Run("summary returns progress", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/games/summary", access, nil)
		if resp.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", resp.status)
		}
		progress := resp.nestedMap("progress")
		if progress["totalRounds"] != float64(1) || progress["totalScore"] != float64(exactScore) {
			t.Fatalf("unexpected progress %v", progress)
		}
	})

	t.Run("rejects score mismatch", func(t *testing.T) {
		round := buildRound("地点B", 30, 100, 31, 101, exactDistance)
		round["score"] = exactScore + 100
		resp := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "classic",
			"totalScore": exactScore + 100,
			"rounds":     []any{round},
		})
		if resp.status != http.StatusBadRequest {
			t.Fatalf("expected 400, got %d", resp.status)
		}
		if !containsMessage(resp, "得分") {
			t.Fatalf("expected score error, got %v", resp.body)
		}
	})

	t.Run("rejects distance mismatch with coordinates", func(t *testing.T) {
		round := buildRound("地点C", 30, 100, 31, 101, 9999)
		round["score"] = 0
		resp := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "classic",
			"totalScore": 0,
			"rounds":     []any{round},
		})
		if resp.status != http.StatusBadRequest {
			t.Fatalf("expected 400, got %d", resp.status)
		}
		if !containsMessage(resp, "距离") {
			t.Fatalf("expected distance error, got %v", resp.body)
		}
	})

	t.Run("rejects total score mismatch", func(t *testing.T) {
		round := buildRound("地点D", 30, 100, 31, 101, exactDistance)
		round["score"] = exactScore
		resp := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "classic",
			"totalScore": 5000,
			"rounds":     []any{round},
		})
		if resp.status != http.StatusBadRequest {
			t.Fatalf("expected 400, got %d", resp.status)
		}
	})

	t.Run("region mode requires region", func(t *testing.T) {
		round := buildRound("地点E", 30, 100, 31, 101, exactDistance)
		round["score"] = exactScore
		resp := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "region",
			"totalScore": exactScore,
			"rounds":     []any{round},
		})
		if resp.status != http.StatusBadRequest {
			t.Fatalf("expected 400, got %d", resp.status)
		}
		if !containsMessage(resp, "region") {
			t.Fatalf("expected region error, got %v", resp.body)
		}
	})

	t.Run("delete removes game", func(t *testing.T) {
		round := buildRound("地点F", 30, 100, 31, 101, exactDistance)
		round["score"] = exactScore
		created := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "classic",
			"totalScore": exactScore,
			"rounds":     []any{round},
		})
		gameID := int64(created.nestedMap("game")["id"].(float64))
		deleted := e.request(t, http.MethodDelete, fmt.Sprintf("/api/games/%d", gameID), access, nil)
		if deleted.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", deleted.status)
		}
		if deleted.body["ok"] != true {
			t.Fatalf("expected ok true, got %v", deleted.body)
		}
		missing := e.request(t, http.MethodDelete, fmt.Sprintf("/api/games/%d", gameID), access, nil)
		if missing.status != http.StatusNotFound {
			t.Fatalf("expected 404, got %d", missing.status)
		}
	})

	t.Run("rejects unauth games", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/games/recent", "", nil)
		if resp.status != http.StatusUnauthorized {
			t.Fatalf("expected 401, got %d", resp.status)
		}
	})
}

func TestDailyChallenge(t *testing.T) {
	e := newTestEnv(t)
	e.seedLocations(30)

	access := registerUser(t, e, uniqueUsername("tester"), randomEmail("daily"), validPassword).
		nestedStr("tokenPair", "accessToken")

	t.Run("today returns 10 locations without answer coords", func(t *testing.T) {
		resp := e.request(t, http.MethodGet, "/api/daily/today", access, nil)
		if resp.status != http.StatusOK {
			t.Fatalf("expected 200, got %d", resp.status)
		}
		locations := resp.nestedArray("locations")
		if len(locations) != 10 {
			t.Fatalf("expected 10 locations, got %d", len(locations))
		}
		if resp.body["played"] != false {
			t.Fatalf("expected played false, got %v", resp.body["played"])
		}
		for _, raw := range locations {
			loc := raw.(map[string]any)
			if _, hasLat := loc["lat"]; hasLat {
				t.Fatal("daily locations must not expose answer coordinates")
			}
		}
	})

	// Fetch today's IDs and guess a round-trip: submit a daily game with a
	// guess point and verify server-authoritative settlement.
	t.Run("submit daily with authoritative settlement", func(t *testing.T) {
		today := e.request(t, http.MethodGet, "/api/daily/today", access, nil)
		locations := today.nestedArray("locations")
		first := locations[0].(map[string]any)
		locationID := first["id"].(float64)

		resp := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "daily",
			"totalScore": 0,
			"rounds": []any{map[string]any{
				"name":       first["name"].(string),
				"locationId": locationID,
				"distanceKm": nil,
				"score":      0,
				"imageId":    nil,
				"xp":         0,
				"difficulty": 1,
				"guessLat":   30.5,
				"guessLng":   100.5,
				"answerLat":  nil,
				"answerLng":  nil,
			}},
		})
		if resp.status != http.StatusCreated {
			t.Fatalf("expected 201, got %d: %v", resp.status, resp.body)
		}
		game := resp.nestedMap("game")
		rounds := resp.nestedArray("game", "rounds")
		storedRound := rounds[0].(map[string]any)
		if storedRound["distanceKm"] == nil || storedRound["answerLat"] == nil {
			t.Fatalf("daily settlement must fill distance and answer coords: %v", storedRound)
		}
		if game["totalScore"] == float64(0) {
			t.Fatal("expected non-zero score for a guess")
		}
	})

	t.Run("second daily submission rejected", func(t *testing.T) {
		today := e.request(t, http.MethodGet, "/api/daily/today", access, nil)
		locations := today.nestedArray("locations")
		first := locations[0].(map[string]any)
		resp := e.request(t, http.MethodPost, "/api/games", access, map[string]any{
			"mode":       "daily",
			"totalScore": 0,
			"rounds": []any{map[string]any{
				"name":       first["name"].(string),
				"locationId": first["id"].(float64),
				"distanceKm": nil,
				"score":      0,
				"imageId":    nil,
				"xp":         0,
				"difficulty": 1,
				"guessLat":   40,
				"guessLng":   110,
				"answerLat":  nil,
				"answerLng":  nil,
			}},
		})
		if resp.status != http.StatusConflict {
			t.Fatalf("expected 409, got %d: %v", resp.status, resp.body)
		}
	})

	t.Run("daily with client-claimed distance rejected", func(t *testing.T) {
		newUser := registerUser(t, e, uniqueUsername("tester"), randomEmail("daily2"), validPassword).
			nestedStr("tokenPair", "accessToken")
		today := e.request(t, http.MethodGet, "/api/daily/today", newUser, nil)
		first := today.nestedArray("locations")[0].(map[string]any)
		resp := e.request(t, http.MethodPost, "/api/games", newUser, map[string]any{
			"mode":       "daily",
			"totalScore": 0,
			"rounds": []any{map[string]any{
				"name":       first["name"].(string),
				"locationId": first["id"].(float64),
				"distanceKm": 10.0,
				"score":      0,
				"imageId":    nil,
				"xp":         0,
				"difficulty": 1,
				"guessLat":   30,
				"guessLng":   100,
				"answerLat":  nil,
				"answerLng":  nil,
			}},
		})
		if resp.status != http.StatusBadRequest {
			t.Fatalf("expected 400, got %d", resp.status)
		}
	})

	t.Run("guest cannot play daily", func(t *testing.T) {
		guest := e.request(t, http.MethodPost, "/api/auth/guest", "", nil)
		today := e.request(t, http.MethodGet, "/api/daily/today", guest.str("guestToken"), nil)
		if today.status != http.StatusOK {
			t.Fatalf("today should be readable for guests, got %d", today.status)
		}
	})
}

func containsMessage(resp *response, fragment string) bool {
	message, ok := resp.body["error"].(string)
	return ok && containsStr(message, fragment)
}

func containsStr(value, fragment string) bool {
	for i := 0; i+len(fragment) <= len(value); i++ {
		if value[i:i+len(fragment)] == fragment {
			return true
		}
	}
	return false
}
